Add support for the Oculus Platform SDK entitlement check

liaxim opened the pull request:
In response to https://github.com/Samsung/GearVRf/issues/808.

GearVRf-DCO-1.0-Signed-off-by: Mihail Marinov <m.marinov@samsung.com>

liaxim commented:
The check in onDrawFrame is a slight concern. Once I know more about this PlatformSDK I will probably move it to the framework's main thread, if deemed necessary.

liaxim commented:
@thomasflynn What should we do about this enhancement? Consider for merging or shelve it?

thomasflynn commented:
@oliverRandom42 has this been helpful to you?  This pull request was generated to help you out.

thomasflynn commented:
this is referencing issue https://github.com/Samsung/GearVRf/issues/808

liaxim commented:
Needs rebase

oliverRandom42 commented:
Hi all,

apologies for slack response, I have been on holiday for the past 2 and a half weeks - will go over this today and provide any feedback.

Cheers,

Oliver

liaxim commented:
Thank you. I am actually re-implementing this to be a framework extension instead. But regardless your feedback is welcome.

liaxim commented:
Rebased and reworked to be a framework extension. To build:

1. In gradle.properties define OVR_PLATFORM_SDK and set it to the location of the platform sdk
2. Go to ./GearVRf/GVRf/Extensions/ and execute ``./gradlew platformsdk_support:assembleDebug``
3. Modify your application to depend on platformsdk_support-debug.aar and do this from onInit

```
        PlatformEntitlementCheck.start(gvrContext, "<your-app-id>", new PlatformEntitlementCheck.ResultListener() {
            @Override
            public void onSuccess() {
                Log.i(TAG, "onSuccess");
            }

            @Override
            public void onFailure() {
                Log.i(TAG, "onFailure");
            }
        });
```

rahul27 commented on GVRf/Extensions/platformsdk_support/src/main/java/org/gearvrf/PlatformEntitlementCheck.java:
```diff
@@ -0,0 +1,100 @@
+     * Starts asynchronous check. Result will be delivered to the listener on the main thread.
+     *
+     * @param context
+     * @param appId application id from
```
Javadoc is incomplete

rahul27 commented on GVRf/Extensions/platformsdk_support/src/main/java/org/gearvrf/PlatformEntitlementCheck.java:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        final GVRActivity activity = context.getActivity();
+        final long result = create(context.getActivity(), appId);
```
can use the activity object from line 58

liaxim commented:
Small tweak: by default do not build the extension: to build have to run this:
```./gradlew -Pplatformsdk_support=true platformsdk_support:assembleDebug```

thomasflynn commented:
@oliverRandom42 let us know if this pull request is helpful

leoncvlt commented:
Hi, I'm working alongside Oliver on the same GearVR App. I compiled the latest library version with `liaxim-oculus_entitlement_check master` changes on top and the ovr_platform_sdk 1.60. Built the extension as instructed, but when testing the above code on the application I get:

```
09-14 16:00:39.996 847-901/com.random42.gvrplayer E/gvrf: ovr_PlatformInitializeAndroid failed with error -1
09-14 16:00:39.996 847-901/com.random42.gvrplayer W/System.err: java.lang.IllegalStateException: Could not initialize the platform sdk; error code: -1
09-14 16:00:39.996 847-901/com.random42.gvrplayer W/System.err:     at org.gearvrf.PlatformEntitlementCheck.start(PlatformEntitlementCheck.java:63)
```

Works fine without executing `PlatformEntitlementCheck.start()`, so I believe I built the framework and extension correctly...

liaxim commented:
@leoncvlt Are you passing a valid app id (the one from the dashboard)? Do you have a build uploaded (in the Build Dashboard)?

liaxim commented:
Having a build uploaded in the dashboard does matter. Unless you have a build (any build would do) the initialization would fail. It is in the platform sdk's documentation.

liaxim commented:
I am pretty sure it doesn't matter but just in case you might try Platform SDK 1.7. The -1 error is pretty hard to figure out. I will re-run my tests and describe what I did.

liaxim commented:
Uploaded a modified sample that works for me: https://github.com/liaxim/GearVRf-Demos/tree/platform_check. I am curious to know whether it would work for you too.

Please send me logs if possible.

liaxim commented:
Rebased

leoncvlt commented:
I tried the gvreyepicking demo form that link (that's the one you added the platform check to, right?). All I did was:

- Added the compiled platformsdk_support-debug and backend-debug libs
- Added the osig for my device
- Replaced the app id string in the `PlatformEntitlementCheck.start` method with my app's from the API section in the oculush dashboard

But still fails with error -1. I'm attaching the logcat.
[logcat.txt](https://github.com/Samsung/GearVRf/files/479864/logcat.txt)


I will try and rebuild the framwork with Platform SDK 1.7 and let you know if that changes anything. Thanks!

liaxim commented:
@leoncvlt Did it work for you with the appId that was specified in there already? I am trying to understand is it something with the implementation or could it be something with your package.

leoncvlt commented:
Leaving the default app id and building the extension and framework with Platform SDK 1.7
still gives me a `ovr_PlatformInitializeAndroid failed with error -1` sadly.

liaxim commented:
Rebased

thomasflynn commented:
@leoncvlt , any progress?

thomasflynn commented:
@rahul27 can you help check this?  thanks.

rahul27 commented:
Will take a look after lunch.

rahul27 commented on GVRf/Extensions/gradle.properties:
```diff
@@ -1,4 +1,9 @@
+#to build the platformsdk-support extension uncomment this property and
+#supply the correct path to the platform sdk. Then run
+#./gradlew -Pplatformsdk_support=true platformsdk_support:assembleDebug
+#OVR_PLATFORM_SDK=<path-to-the-oculus-platform-sdk
```
close angle brackets

liaxim replied:
onFailure is supposed to be called in case the actual entitlement check fails. If you are in no position to even submit the request you get the exception. It is mentioned in the javadoc.

liaxim replied:
Missing bracket added

rahul27 commented:
@liaxim I see the same error as @leoncvlt 

```
09-28 10:18:02.991 20299-20343/org.gearvrf.simplesample E/gvrf: ovr_PlatformInitializeAndroid failed with error -1
09-28 10:18:02.991 20299-20343/org.gearvrf.simplesample W/System.err: java.lang.IllegalStateException: Could not initialize the platform sdk; error code: -1
09-28 10:18:02.991 20299-20343/org.gearvrf.simplesample W/System.err:     at org.gearvrf.PlatformEntitlementCheck.start(PlatformEntitlementCheck.java:63)
09-28 10:18:02.991 20299-20343/org.gearvrf.simplesample W/System.err:     at org.gearvrf.simplesample.SampleMain.onInit(SampleMain.java:66)
09-28 10:18:02.991 20299-20343/org.gearvrf.simplesample W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
09-28 10:18:02.991 20299-20343/org.gearvrf.simplesample W/System.err:     at org.gearvrf.GVREventManager.invokeMethod(GVREventManager.java:307)
```

Am using the app id given to me from the oculus dashboard. I was expecting an onFailure() since I do not have a build uploaded. 

See the same with the app id from your commit.

liaxim commented:
So I re-did the whole thing. Even though my new submission checked out according to the validator I was getting -1 too.

This is how I got it to work. Run the phone not in developer mode, dock, make sure you get to the Oculus home. Launch the app then (debug build now) but still not in developer mode. Dock and check the result of the query - this is when it started working for me. Afterwards it worked in developer mode too.

Some steps might be optional but this is what I did. There is nothing we can do on our side to improve this very finicky process. We will get it to work together tomorrow.

I used the 1.8.0 platform sdk but based on my past experience 1.6 and 1.7 work fine too.

rahul27 commented:
@liaxim took a lot of trying but I finally got it to work.

rahul27 commented:
This one is good to go.

thomasflynn commented:
@leoncvlt do the steps @liaxim mentions above help?

leoncvlt commented:
Apologies for the lack of participation, I will run some tests tomorrow and let you know

thomasflynn commented:
@leoncvlt ?

liaxim commented:
Just modify the manifest of your to app to specify a category of android.intent.category.LAUNCHER. Have developer mode disabled. Launch your app, dock.

Beforehand make sure you can get to Oculus home without needing updates on docking the phone. Like dock while in the system launcher, do you get Oculus home without any prompts? I recommend you sign in with your Oculus account too.

liaxim commented:
You may find this useful though it is not exactly on topic: https://forums.oculus.com/developer/discussion/40097/platform-sdk-1-5-0-cant-initialize-platform.

leoncvlt commented:
Cheers for the link Liaxim, cleared a few things up. Going forward! Now the app doesn't crash, but PlatformEntitlementCheck doesn't return a success either - I get this exception on the start() method:

```
W/System.err: android.content.pm.PackageManager$NameNotFoundException: com.oculus.shimwow
W/System.err:     at android.app.ApplicationPackageManager.getPackageInfo(ApplicationPackageManager.java:201)
W/System.err:     at org.gearvrf.PlatformEntitlementCheck.create(Native Method)
W/System.err:     at org.gearvrf.PlatformEntitlementCheck.start(PlatformEntitlementCheck.java:61)
W/System.err:     at com.random42.gvrshowreel.GVRVideo.onInit(GVRVideo.java:56)
W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
```

liaxim commented:
Hi, could you share the full logcat output? I have seen this exception before but it got cleared after going to Oculus home then launching the app with developer mode off from the regular android launcher.

leoncvlt commented:
Just tried again, uninstalled the app and started fresh making sure oculus home was able to run fine and developer mode was off. Finally got a PlatformEntitlementCheck Success!

liaxim commented:
Awesome. I wish there was something we could do on our side to simplify this.

leoncvlt commented:
No worries, thanks for all your help and patience so far.

thomasflynn commented:
@rahul27 , i think you previously volunteered to document this process?  we should definitely document this and add it to our website.  thanks.

rahul27 commented:
Yup. Will add a page to the wiki.

rahul27 commented:
Here is a link to the wiki:

https://github.com/Samsung/GearVRf/wiki/Entitlement-Check-using-GVRF
